### ISSUE TITLE
[Feature] Provide an easy way to configure logging for a job

### ISSUE THREAD

**Issue by chhabrakadabra** [ray-project/ray#23385]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

If I configure `logging` in my driver script, that doesn't translate to my remote task's logging being configured the same way. Today, I have to configure my logging inside my remote tasks/actors too. If I use a library that uses Ray primitives (including builtin abstractions like ray datasets), I have no ability to control the logging configs inside the ray tasks created by those libraries.

It would be nice if there was some way of configuring logging that would automatically be applied to all the remote tasks.

### Use case

We use centralized logging that parses JSON encoded log messages. To take full advantage of this, the logs we output should be JSON formatted. We want to make it easy for people to configure logging and doing it once from the driver would be much better than assuming that people remember to do it from every remote task.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by krfricke** [comment]
I agree - not being able to configure logging globally produces unavoidable log spam in our libraries.

For instance, Ray Serve logs every incoming request with INFO level. It's currently impossible to disable this, as driver settings are ignored and we can't inject logic into the serving workers.

Simple repro:

```
import logging
import ray

ray.init()


sub = logging.getLogger("sub")
sub.setLevel(logging.ERROR)

sub.error("This is visible")
sub.warning("This is not visible")


@ray.remote
def log_in_process():
    sub.error("This is visible (and should be)")
    sub.warning("This is visible (but should _not_ be)")


ray.get(log_in_process.remote())
```

**Comment by scottsun94** [comment]
cc: @rkooo567

**Comment by trahloff** [comment]
Hi ray team. Is this on the roadmap?

**Comment by rkooo567** [comment]
yes! From Ray 2.6, you should be able to use the experimental `worker_process_setup_hook` API. https://docs.ray.io/en/master/ray-observability/user-guides/configure-logging.html#customizing-worker-process-loggers

It has a couple issues now (e.g., it doesn't work with ray job submission). We are planning to handle it in 2.7

**Comment by rkooo567** [comment]
Please try the API!

**Comment by jjyao** [comment]
cc @kevin85421

**Comment by jonasdedden** [comment]
Hey, I stumbled over this issue and it unfortunately makes using Ray a whole lot more cumbersone than initially thought.

I am in a context where I use Ray *Client* + KubeRay in k8s, so unfortunately the `worker_process_setup_hook` approach doesn't work for me: https://docs.ray.io/en/latest/ray-observability/user-guides/configure-logging.html#customizing-worker-process-loggers

> This is an experimental feature. The semantic of the API is subject to change. It doesn’t support [Ray Client](https://docs.ray.io/en/latest/cluster/running-applications/job-submission/ray-client.html#ray-client-ref) yet.

I'm using this test method:
```Python
def do_work(x):
    import sys
    print(f"stdout {x} test", file=sys.stdout)
    print(f"stderr {x} test", file=sys.stderr)
    LOGGER.debug(f"logging debug {x} test")
    LOGGER.info(f"logging info {x} test")
    LOGGER.warning(f"logging warning {x} test")
    LOGGER.error(f"logging error {x} test")
    return x*x
```

I tried calling this function whilst supplying a proper logging initialization function to `worker_process_setup_hook` in the `ray.init` method, which didn't do anything, but even if I supply a total gibberish string there,

```Python
ray.init("ray://[address somewhere in k8s]:10001", log_to_driver=False, runtime_env={"worker_process_setup_hook": "brlgh"})
```

I don't even get an error message anywhere. So I really think that the setup hook is not called at all in Ray Client contexts.


I even tried out the not publicy documented `RuntimeEnvPlugin`s:

```Python
# foobar/__init__.py
from typing import List, Optional

from ray._private.runtime_env.context import RuntimeEnvContext
from ray._private.runtime_env.plugin import RuntimeEnvPlugin

import logging

class LoggingPlugin(RuntimeEnvPlugin):
    name: str = "LoggingPlugin"
    
    async def create(
        self,
        uri: Optional[str],
        runtime_env,
        context: RuntimeEnvContext,
        logger: logging.Logger,
    ) -> float:
        """Create and install the runtime environment.

        Gets called in the runtime env agent at install time. The URI can be
        used as a caching mechanism.

        Args:
            uri: A URI uniquely describing this resource.
            runtime_env: The RuntimeEnv object.
            context: Auxiliary information supplied by Ray.
            logger: A logger to log messages during the context modification.

        Returns:
            float: The disk space taken up by this plugin installation for this
                environment. e.g. for working_dir, this downloads the files to the
                local node.
        """
        import logging
        logging.basicConfig(level=logging.DEBUG)
        logger.error("I AM IN CREATE!")
        print("STDOUT I AM IN CREATE!")
        return 0

    def modify_context(
        self,
        uris: List[str],
        runtime_env: "RuntimeEnv",  # noqa: F821
        context: RuntimeEnvContext,
        logger: logging.Logger,
    ) -> None:
        """Modify context to change worker startup behavior.

        For example, you can use this to prepend "cd <dir>" command to worker
        startup, or add new environment variables.

        Args:
            uris: The URIs used by this resource.
            runtime_env: The RuntimeEnv object.
            context: Auxiliary information supplied by Ray.
            logger: A logger to log messages during the context modification.
        """
        logging.basicConfig(level=logging.DEBUG)
        logger.error("I AM IN MODIFY_CONTEXT!")
        print("STDOUT I AM IN MODIFY_CONTEXT!")
        return
```

With the environment variable `RAY_RUNTIME_ENV_PLUGINS=[{"class":"foobar.LoggingPlugin"}]` on the workers and using the Ray Client code,
```Python
ray.init("ray://[address in ks]:10001", log_to_driver=False, runtime_env={"LoggingPlugin": {}})
wrapped_func = ray.remote(do_work).options(runtime_env={"LoggingPlugin": {}})  # I don't think it's required here again
wrapped_func.remote(...)
```
I *do* get the messages in the workers logs from the logging & print calls I placed in `modify_context` & `create` of the *Plugin*,
```
(base) ray@6fa2cccef0fe:/tmp/ray/session_latest/logs$ grep -R "I AM" *
runtime_env_agent.out:STDOUT I AM IN CREATE!
runtime_env_agent.out:STDOUT I AM IN MODIFY_CONTEXT!
runtime_env_agent.out:STDOUT I AM IN CREATE!
runtime_env_agent.out:STDOUT I AM IN MODIFY_CONTEXT!
runtime_env_setup-01000000.log:2025-11-06 07:34:14,951	ERROR __init__.py:36 -- I AM IN CREATE!
runtime_env_setup-01000000.log:2025-11-06 07:34:14,951	ERROR __init__.py:59 -- I AM IN MODIFY_CONTEXT!
runtime_env_setup-01000000.log:2025-11-06 07:34:15,155	ERROR __init__.py:36 -- I AM IN CREATE!
runtime_env_setup-01000000.log:2025-11-06 07:34:15,155	ERROR __init__.py:59 -- I AM IN MODIFY_CONTEXT!
```
but the logging level of the actual Ray worker process seems unchanged, i.e. logging INFO level messages of `do_work` are nowhere to be seen. But I'm *very* sure I'm misusing the RuntimePlugin interfaces here anyways 😅 

I also collected other issues I have with Ray logging in the support Slack workspace: https://ray.slack.com/archives/C02GFQ82JPM/p1762276075972359

**Comment by jonasdedden** [comment]
Okay, I found a solution which borders on being absolutely crazy 😅 

I found that there is the definitely undocumented environment variable "`__RAY_WORKER_PROCESS_SETUP_HOOK_ENV_VAR`" (https://github.com/ray-project/ray/blob/9c0dc3ce6b09673caad4fd38c6939b644381efd2/python/ray/_private/ray_constants.py#L516C38-L516C77) which actually is used exactly in the same way as supplying `worker_process_setup_hook` in the worker setup: https://github.com/ray-project/ray/blob/9c0dc3ce6b09673caad4fd38c6939b644381efd2/python/ray/_private/workers/default_worker.py#L314 and https://github.com/ray-project/ray/blob/9c0dc3ce6b09673caad4fd38c6939b644381efd2/python/ray/_private/runtime_env/setup_hook.py#L140

So basically, one can circumvent that `worker_process_setup_hook` does not work in Ray Client / static KubeRay environments by supplying the `__RAY_WORKER_PROCESS_SETUP_HOOK_ENV_VAR` env var statically in your k8s Worker pods with the name of a function that contains your specific logging setup. 🚀 😅 

I tested this, now logging levels which I expected are correctly coming out.

This either should probably be way better documented, or the interface for init hooks in `ray.init(...)` should be fixed in Ray Client environments. I also found another logging setting issue which appears to be exclusively happening in Ray Client contexts (https://github.com/ray-project/ray/issues/58430).
